osu.Game.Rulesets.Osu/UI/OsuPlayfield.cs

```diff
@@ -92,6 +128,67 @@ private void onNewResult(DrawableHitObject judgedObject, JudgementResult result)
             judgementLayer.Add(explosion);
         }
 
+        /// <summary>
+        /// Misses all <see cref="OsuHitObject"/>s occurring earlier than the start time of a judged <see cref="OsuHitObject"/>.
+        /// </summary>
+        /// <param name="result">The <see cref="JudgementResult"/> of the judged <see cref="OsuHitObject"/>.</param>
+        private void missAllEarlier(JudgementResult result)
+        {
+            if (!causesNoteLockMisses(result.HitObject))
+                return;
+
+            // The minimum start time required for hitobjects so that they aren't missed.
+            double minimumTime = result.HitObject.StartTime;
+
+            foreach (var obj in HitObjectContainer.AliveObjects)
+            {
+                if (obj.HitObject.StartTime >= minimumTime)
+                    break;
+
+                performMiss(obj);
+
+                foreach (var n in obj.NestedHitObjects)
+                {
+                    if (n.HitObject.StartTime >= minimumTime)
+                        break;
+
+                    performMiss(n);
+                }
+            }
+        }
+
+        private void performMiss(DrawableHitObject obj)
+        {
+            if (!(obj is DrawableOsuHitObject osuObject))
+                throw new InvalidOperationException($"{obj.GetType()} is not a {nameof(DrawableOsuHitObject)}.");
+
+            // Hitobjects that have already been judged cannot be missed.
+            if (osuObject.Judged)
+                return;
+
+            if (!causesNoteLockMisses(obj.HitObject))
+                return;
+
+            osuObject.MissForcefully();
+        }
+
+        /// <summary>
+        /// Whether a <see cref="HitObject"/> is contributes to note lock.
+        /// Future contributing <see cref="HitObject"/>s will not be hittable until the start time of the last contributing <see cref="HitObject"/> is reached.
+        /// </summary>
+        /// <param name="hitObject">The <see cref="HitObject"/> to test.</param>
+        /// <returns>Whether <paramref name="hitObject"/> causes note lock.</returns>
+        private bool contributesToNoteLock(HitObject hitObject)
+            => hitObject is HitCircle || hitObject is Slider;
```

Comment:
curious why this is matching on `Slider` but the one below is matching on `SliderTailCircle`. logically i'd expect this to be checking against `SliderHeadCircle`, not `Slider` itself.

Reply:
Because `SliderHeadCircle` isn't a part of the hitobjects list, whereas it _is_ apart of the judgements. Maybe I should comment that?

Reply:
it is in nested though right? maybe i misread the logic but it looked like you were also checking nested objects.

Reply:
I'm only checking nested for the miss loop which happens on judgement. This is a separate loop that happens every `CheckHittable`. I could check nested in both cases, but with the added complexity I'm not sure if it's really worth it... will think about it.